### ISSUE TITLE
Migrating to Micronaut 2.0 - Could not obtain transaction-synchronized Session for current thread

### ISSUE THREAD

**Issue by sdelamo** [micronaut-projects/micronaut-sql#130]
I am trying to Migrate to Micronaut 2.0.0.M1 a Micronaut  1.3.3 application with following dependencies: 

```groovy
dependencies {
    ...
    annotationProcessor "javax.persistence:javax.persistence-api:2.2"
    implementation "io.micronaut.configuration:micronaut-hibernate-jpa"
    implementation "io.micronaut.configuration:micronaut-liquibase"
    implementation "io.micronaut.configuration:micronaut-jdbc-tomcat"
    runtimeOnly "mysql:mysql-connector-java:8.0.15"
}
    
```

I have this kind of repository:

```java
package example.micronaut;

import example.micronaut.entities.BookEntity;
import io.micronaut.configuration.hibernate.jpa.scope.CurrentSession;
import io.micronaut.spring.tx.annotation.Transactional;
import javax.annotation.Nonnull;
import javax.inject.Singleton;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.Query;
import javax.persistence.TypedQuery;
import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaDelete;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;

@Singleton
public class DefaultBookRepository implements BookRepository {

    @PersistenceContext
    private final EntityManager entityManager;

    /**
     * Instantiates a RequestFingerPrintRepositoryImpl.
     *
     * @param entityManager The entity manager
     */
    public DefaultBookRepository(@CurrentSession EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    @Transactional
    @Override
    public Long save(@Nonnull @NotBlank String name) {
        BookEntity entity = new BookEntity();
        entity.setName(name);
        entityManager.persist(entity);
        return entity.getId();
    }

    @Transactional(readOnly = true)
    @Nonnull
    @Override
    public Number count() {
        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        CriteriaQuery<Long> cr = builder.createQuery(Long.class);
        Root<BookEntity> root = cr.from(BookEntity.class);
        cr.select(builder.count(root.get("id")));
        TypedQuery<Long> query = entityManager.createQuery(cr);
        return query.getSingleResult();
    }

    @Transactional
    @Override
    public void deleteById(@Nonnull @NotNull Long id) {
        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        CriteriaDelete<BookEntity> cr = builder.createCriteriaDelete(BookEntity.class);
        Root<BookEntity> root = cr.from(BookEntity.class);
        cr.where(builder.equal(root.get("id"), id));
        Query query = entityManager.createQuery(cr);
        query.executeUpdate();
    }
}
```

## Migration Steps

I set micronaut version to 2.0.0.M1 and I took the following steps:

### Remove @CurrentSession

I removed `@CurrentSession` since it is deprecated.

```
@deprecated It is no longer necessary to specify {@code CurrentSession} to obtain an entity manager
```

### Resolve Transactional Import

Import `import io.micronaut.spring.tx.annotation.Transactional;` could not longer be resolved. 

I added `implementation "io.micronaut:micronaut-spring"` to resolve it.

### Replace deprecated io.micronaut.spring.tx.annotation.Transactional

`io.micronaut.spring.tx.annotation.Transactional` is deprecated. I replace it with the suggestion: 

`* @deprecated Use Spring's {@link org.springframework.transaction.annotation.Transactional} directly or switch to Micronaut Data's {@code TransactionalAdvice}`

Thus, I replaced

`io.micronaut.spring.tx.annotation.Transactional`

with:

`org.springframework.transaction.annotation.Transactional`

### Add a transaction manager

If I run the tests I get:

```
No transaction manager configured
org.springframework.transaction.TransactionSystemException: No transaction manager configured
	at io.micronaut.spring.tx.annotation.TransactionInterceptor.resolveTransactionManager(TransactionInterceptor.java:143)
	at io.micronaut.spring.tx.annotation.TransactionInterceptor.intercept(TransactionInterceptor.java:66)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at example.micronaut.BookRepositorySpec.it is possible to save a book(BookRepositorySpec.groovy:10)
```

[Spring Data Support](https://micronaut-projects.github.io/micronaut-data/snapshot/guide/index.html#spring) section of Micronaut Data docs states:

> In addition to this dependency you will need either spring-orm (for Hibernate) or spring-jdbc (for JDBC) on your classpath to enable support for Spring-based transaction management:

I added: 

`implementation("org.springframework:spring-orm:5.2.0.RELEASE")`

### Could not obtain transaction-synchronized Session for current thread

Now I get: 

```
ould not obtain transaction-synchronized Session for current thread
org.hibernate.HibernateException: Could not obtain transaction-synchronized Session for current thread
	at org.springframework.orm.hibernate5.SpringSessionContext.currentSession(SpringSessionContext.java:143)
	at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:473)
	at io.micronaut.configuration.hibernate.jpa.TransactionalSessionInterceptor.intercept(TransactionalSessionInterceptor.java:56)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at example.micronaut.DefaultBookRepository.count(DefaultBookRepository.java:47)
	at example.micronaut.BookRepositorySpec.it is possible to save a book(BookRepositorySpec.groovy:10)
```

# Example Repository: 

https://github.com/grails-core-issues-forks/micronaut-sql-130

Steps to reproduce:

Clone the repository:

## Micronaut 1.3. 

```
% cd micronaut13
micronaut13 % ./gradlew test
 BUILD SUCCESSFUL
```

## Micronaut 2.0

```
% cd micronaut20
micronaut20 % ./gradlew test
> Task :test FAILED
```

**Comment by graemerocher** [comment]
I think the issue may be that Micronaut Data currently maps `Transactional` for spring but it is missing from this project, so for the meantime you would have to use the deprecated `io.micronaut.spring.tx.annotation.Transactional`

**Comment by graemerocher** [comment]
Or add `micronaut-spring-annotation` to your annotation processor classpath, which also has a mapper https://github.com/micronaut-projects/micronaut-spring/blob/master/spring-annotation/src/main/java/io/micronaut/spring/annotation/tx/TransactionalAnnotationMapper.java

**Comment by sdelamo** [comment]
If I use the depreacated `io.micronaut.spring.tx.annotation.Transactional` I get: 

```
no transaction is in progress
javax.persistence.TransactionRequiredException: no transaction is in progress
	at org.hibernate.internal.AbstractSharedSessionContract.checkTransactionNeededForUpdateOperation(AbstractSharedSessionContract.java:413)
	at org.hibernate.internal.SessionImpl.checkTransactionNeededForUpdateOperation(SessionImpl.java:3387)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1343)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1339)
	at org.springframework.orm.hibernate5.SessionFactoryUtils.flush(SessionFactoryUtils.java:148)
	at org.springframework.orm.hibernate5.SpringSessionSynchronization.beforeCommit(SpringSessionSynchronization.java:95)
	at org.springframework.transaction.support.TransactionSynchronizationUtils.triggerBeforeCommit(TransactionSynchronizationUtils.java:96)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.triggerBeforeCommit(AbstractPlatformTransactionManager.java:920)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:728)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:712)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:631)
	at io.micronaut.spring.tx.annotation.TransactionInterceptor.intercept(TransactionInterceptor.java:88)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at io.micronaut.validation.ValidatingInterceptor.intercept(ValidatingInterceptor.java:108)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:69)
	at example.micronaut.BookRepositorySpec.it is possible to save a book(BookRepositorySpec.groovy:7)
```

**Comment by sdelamo** [comment]
I've tried adding `annotationProcessor "io.micronaut.spring:micronaut-spring-annotation"` but I get the same `javax.persistence.TransactionRequiredException: no transaction is in progress` exception.

**Comment by graemerocher** [comment]
Problem can be resolved with the following diff:

```diff
From c03b2d7282767a9831b4f7eb968ea25108a3704c Mon Sep 17 00:00:00 2001
From: graemerocher <graeme.rocher@gmail.com>
Date: Tue, 14 Apr 2020 09:51:31 +0200
Subject: [PATCH 1/1] Fix Micronaut 2.0 project

---
 micronaut20/build.gradle                                     | 5 +++--
 micronaut20/gradle.properties                                | 2 +-
 .../main/java/example/micronaut/DefaultBookRepository.java   | 2 --
 3 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/micronaut20/build.gradle b/micronaut20/build.gradle
index 8af10e4..9f7443d 100644
--- a/micronaut20/build.gradle
+++ b/micronaut20/build.gradle
@@ -9,7 +9,6 @@ version "0.1"
 group "example.micronaut"

 repositories {
-    mavenCentral()
     maven { url "https://jcenter.bintray.com" }
 }

@@ -22,6 +21,7 @@ dependencies {
     annotationProcessor platform("io.micronaut:micronaut-bom:$micronautVersion")
     annotationProcessor "io.micronaut:micronaut-inject-java"
     annotationProcessor "io.micronaut:micronaut-validation"
+    annotationProcessor "io.micronaut.spring:micronaut-spring-annotation"
     implementation platform("io.micronaut:micronaut-bom:$micronautVersion")
     implementation "io.micronaut:micronaut-inject"
     implementation "io.micronaut:micronaut-validation"
@@ -32,10 +32,11 @@ dependencies {

     annotationProcessor "javax.persistence:javax.persistence-api:2.2"
     implementation "io.micronaut.configuration:micronaut-hibernate-jpa"
+    implementation "io.micronaut.configuration:micronaut-hibernate-jpa-spring:2.0.0.RC1"
     implementation "io.micronaut.configuration:micronaut-liquibase"
     implementation "io.micronaut.configuration:micronaut-jdbc-tomcat"
     runtimeOnly "mysql:mysql-connector-java:8.0.15"
-    implementation "io.micronaut:micronaut-spring"
+    implementation "io.micronaut:micronaut-spring:2.0.1"
     implementation("org.springframework:spring-orm:5.2.0.RELEASE")

     testAnnotationProcessor platform("io.micronaut:micronaut-bom:$micronautVersion")
diff --git a/micronaut20/gradle.properties b/micronaut20/gradle.properties
index e024f11..c87cccd 100644
--- a/micronaut20/gradle.properties
+++ b/micronaut20/gradle.properties
@@ -1,2 +1,2 @@
-micronautVersion=2.0.0.M1
+micronautVersion=2.0.0.M2
 testContainersVersion=1.12.3
diff --git a/micronaut20/src/main/java/example/micronaut/DefaultBookRepository.java b/micronaut20/src/main/java/example/micronaut/DefaultBookRepository.java
index 31abe4c..61ea30f 100644
--- a/micronaut20/src/main/java/example/micronaut/DefaultBookRepository.java
+++ b/micronaut20/src/main/java/example/micronaut/DefaultBookRepository.java
@@ -6,7 +6,6 @@ import org.springframework.transaction.annotation.Transactional;
 import javax.annotation.Nonnull;
 import javax.inject.Singleton;
 import javax.persistence.EntityManager;
-import javax.persistence.PersistenceContext;
 import javax.persistence.Query;
 import javax.persistence.TypedQuery;
 import javax.persistence.criteria.CriteriaBuilder;
@@ -19,7 +18,6 @@ import javax.validation.constraints.NotNull;
 @Singleton
 public class DefaultBookRepository implements BookRepository {

-    @PersistenceContext
     private final EntityManager entityManager;

     /**
--
2.24.1 (Apple Git-126)
```

**Comment by sdelamo** [comment]
I added a commit to the sample repo updating the app in case anyone needs to perform a similar update.
